nx/lib/nx/lin_alg/svd.ex

```diff
@@ -0,0 +1,251 @@
+defmodule Nx.LinAlg.SVD do
+  import Nx.Defn
+  @eps 1.1920929e-07
+
+  defn svd(input_tensor, opts \\ []) do
+    validate_opts(opts)
+
+    {is_flipped, tensor} =
+      case Nx.shape(input_tensor) do
+        {m, n} when m < n ->
+          {true, Nx.LinAlg.adjoint(input_tensor)}
+
+        _ ->
+          {false, input_tensor}
+      end
+
+    # We always return full matrices for retrocompatibility.
+    # Original Jax code has extensions if we want to add that
+    # as an option (i.e. mode: :complete | :reduced)
+
+    {reduce_to_square, q, u_null, a} =
+      case Nx.shape(tensor) do
+        {m, n} when m > n ->
+          {q_full, a_full} = Nx.LinAlg.qr(tensor, mode: :complete)
+          q = q_full[[0..-1//1, 0..(n - 1)//1]]
+          u_null = q_full[[0..-1//1, n..-1//1]]
+          a = a_full[0..(n - 1)//1]
+          {true, q, u_null, a}
+
+        {n, n} ->
+          {false, tensor, tensor, tensor}
+      end
+
+    {u, s, v} = svd_tall_and_square(a, opts)
+
+    u =
+      case reduce_to_square do
+        true ->
+          u = Nx.dot(q, u)
+          Nx.concatenate([u, u_null], axis: -1)
+
+        false ->
+          u
+      end
+
+    res =
+      case is_flipped do
+        true -> {v, s, Nx.LinAlg.adjoint(u)}
+        false -> {u, s, Nx.LinAlg.adjoint(v)}
+      end
+
+    custom_grad(res, [input_tensor], fn g ->
+      grad(res, input_tensor, g)
+    end)
+  end
+
+  deftransformp validate_opts(opts \\ []) do
+    opts[:max_iter] || raise ArgumentError, "missing option :max_iter"
+  end
+
+  defnp svd_tall_and_square(a, opts \\ []) do
+    {_m, n} = Nx.shape(a)
+    {u, h} = qdwh(a, opts)
+    # ensure H is hermitian
+    h = (h + Nx.LinAlg.adjoint(h)) / 2
+    {s, v} = Nx.LinAlg.eigh(h, max_iter: opts[:max_iter])
+
+    sign = Nx.select(s < 0, -1, 1)
+    v = sign * v
+    s = sign * s
+
+    # sort s and v according to
+    sort_idx = Nx.argsort(s, direction: :desc)
+    s_out = Nx.take(s, sort_idx)
+    v_out = Nx.take(v, sort_idx, axis: 1)
+    u_out = Nx.dot(u, v_out)
+
+    u_out = Nx.select(s[0] < n * @eps * s_out[0], correct_rank_deficiency(u_out), u_out)
+    {u_out, s_out, v_out}
+  end
+
+  defn qdwh(x, opts \\ []) do
+    alpha = Nx.sqrt(Nx.LinAlg.norm(x, ord: 1)) * Nx.sqrt(Nx.LinAlg.norm(x, ord: :inf))
+    l = @eps
+    u = x / alpha
+    tol_l = 5 * @eps
+    tol_norm = Nx.cbrt(tol_l)
+
+    one_u8 = Nx.tensor(1, type: :u8)
+
+    {u, _l, _num_iters, _is_unconverged, _is_not_max_iteration, _max_iter} =
+      while {u, l, iter_idx = 1, is_unconverged = one_u8, is_not_max_iteration = one_u8,
```

Comment:
```suggestion
    {u, _l, _num_iters, _is_unconverged, _is_not_max_iteration, _max_iter} =
      while {u, l, iter_idx = 1, is_unconverged = true, is_not_max_iteration = one_u8,
```

Reply:
I didn't know this worked :)

Reply:
Didn't work :(

Reply:
Ups sorry, I forgot that when a boolean is passed in opts it's cast to :u8